cmd/virt-handler/virt-handler.go

```diff
@@ -275,18 +289,36 @@ func (app *virtHandlerApp) Run() {
 	stop := make(chan struct{})
 	defer close(stop)
 	factory.Start(stop)
-	cache.WaitForCacheSync(stop, factory.ConfigMap().HasSynced)
+	cache.WaitForCacheSync(stop, factory.ConfigMap().HasSynced, vmiInformer.HasSynced)
 
 	go vmController.Run(10, stop)
 
-	logger.V(1).Infof("metrics: max concurrent requests=%d", app.MaxRequestsInFlight)
+	errCh := make(chan error)
+	go app.runPrometheusServer(errCh, certStore)
+	go app.runConsoleServer(errCh, clusterConfig.GetConsoleServerPort(), consoleHandler)
+
+	// wait for one of the servers to exit
+	<-errCh
+}
+
+func (app *virtHandlerApp) runPrometheusServer(errCh chan error, certStore certificate.FileStore) {
+	log.Log.V(1).Infof("metrics: max concurrent requests=%d", app.MaxRequestsInFlight)
 	http.Handle("/metrics", promvm.Handler(app.MaxRequestsInFlight))
+	errCh <- http.ListenAndServeTLS(app.ServiceListen.Address(), certStore.CurrentPath(), certStore.CurrentPath(), nil)
+}
 
-	err = http.ListenAndServeTLS(app.ServiceListen.Address(), certStore.CurrentPath(), certStore.CurrentPath(), nil)
-	if err != nil {
-		log.Log.Reason(err).Error("Serving prometheus failed.")
-		panic(err)
+func (app *virtHandlerApp) runConsoleServer(errCh chan error, port int, consoleHandler *rest.ConsoleHandler) {
+	ws := new(restful.WebService)
+	ws.Route(ws.GET("/namespaces/{namespace}/vmi/{name}/console").To(consoleHandler.SerialHandler))
```

Comment:
Can we use `{namespace}/virtualmachineinstances/{name}` like on the k8s apiserver?

Reply:
concretely, you mean to rename `vmi` to `virtualmachineinstances` ?

Reply:
Yep. Also forgot to mention that we should have a version prefix. `v1`. Probably (not related to our public api).